mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -251,12 +253,72 @@ def _list_experiments(
         query_options = self._get_eager_experiment_query_options() if eager else []
         return session.query(SqlExperiment).options(*query_options).filter(*conditions).all()
 
-    def list_experiments(self, view_type=ViewType.ACTIVE_ONLY):
+    def _list_experiments(
+        self,
+        ids=None,
+        names=None,
+        view_type=ViewType.ACTIVE_ONLY,
+        max_results=None,
+        page_token=None,
+        eager=False,
+    ):
+        """
+        :param max_results: If passed, specifies the maximum number of experiments desired. If not
+                            passed, all experiments will be returned.
+        :param page_token: Token specifying the next page of results. It should be obtained from
+                            a ``list_experiments`` call.
+        :param eager: If ``True``, eagerly loads each experiments's tags. If ``False``, these tags
+                      are not eagerly loaded and will be loaded if/when their corresponding
+                      object properties are accessed from a resulting ``SqlExperiment`` object.
+        """
+        stages = LifecycleStage.view_type_to_stages(view_type)
+        conditions = [SqlExperiment.lifecycle_stage.in_(stages)]
+        if ids and len(ids) > 0:
+            int_ids = [int(eid) for eid in ids]
+            conditions.append(SqlExperiment.experiment_id.in_(int_ids))
+        if names and len(names) > 0:
+            conditions.append(SqlExperiment.name.in_(names))
+
+        max_results_for_query = None
+        if max_results is not None:
+            max_results_for_query = max_results + 1
+
+            def compute_next_token(current_size):
+                next_token = None
+                if max_results_for_query == current_size:
+                    final_offset = offset + max_results
+                    next_token = SearchUtils.create_page_token(final_offset)
+
+                return next_token
+
         with self.ManagedSessionMaker() as session:
-            return [
-                exp.to_mlflow_entity()
-                for exp in self._list_experiments(session=session, view_type=view_type, eager=True)
-            ]
+            query_options = self._get_eager_experiment_query_options() if eager else []
+            if max_results is not None:
+                offset = SearchUtils.parse_start_offset_from_page_token(page_token)
+                queried_experiments = (
+                    session.query(SqlExperiment)
+                    .options(*query_options)
+                    .filter(*conditions)
+                    .offset(offset)
+                    .limit(max_results_for_query)
+                    .all()
```

Comment:
Do we need to sort experiments (by experiment name and id) here (note there is creation_time column in open-source MLflow)?

Reply:
Thanks for making all of these changes @harupy. I don't see any guarantee currently about experiments being sorted, so I don't think that's a requirement for this feature. It might be a useful follow-up item though.

Reply:
Agree that this is a useful follow-up item. May we could sort experiments by `experiment_id` in decsending order so newer experiments are fetched early.